spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/support/SimpleNeo4jRepository.java

```diff
@@ -190,24 +189,9 @@ public void deleteAll() {
 
 	@Override
 	public Page<T> findAll(Pageable pageable, int depth) {
-		Collection<T> data = session.loadAll(clazz, PagingAndSortingUtils.convert(pageable.getSort())
-				, new Pagination(pageable.getPageNumber(), pageable.getPageSize()), depth);
-		return updatePage(pageable, new ArrayList<>(data));
-	}
-
-	/*
-	 * This is a cheap trick to estimate the total number of objects without actually knowing the real value.
-	 * Essentially, if the result size is the same as the page size, we assume more data can be fetched, so
-	 * we set the expected total to the current total retrieved so far + the current page size. As soon as the
-	 * result size is less than the page size, we know there are no more, so we set the total to the number
-	 * retrieved so far. This will ensure that page.next() returns false.
-	 */
-	private Page<T> updatePage(Pageable pageable, List<T> results) {
-
-		int pageSize = pageable.getPageSize();
-		long pageOffset = pageable.getOffset();
-		long total = pageOffset + results.size() + (results.size() == pageSize ? pageSize : 0);
+		Pagination pagination = new Pagination(pageable.getPageNumber(), pageable.getPageSize());
+		Collection<T> data = session.loadAll(clazz, PagingAndSortingUtils.convert(pageable.getSort()), pagination, depth);
 
-		return new PageImpl<T>(results, pageable, total);
+		return PageableExecutionUtils.getPage(new ArrayList<>(data), pageable, () -> session.countEntitiesOfType(clazz));
```

Comment:
To bad that `getPage` needs a list, not a collection. I just checked, we return a `Set` and so we are copying stuff around.

Reply:
There will be no API change in SD commons in the near future. As far as we checked this `Collection` is already an `ArrayList` but I won't do a downcast here.

Reply:
Fully understandable.